### PR TITLE
First draft of MUSDB18 dataset

### PR DESCRIPTION
**Pull request by faroit**
this PR adds the MUSDB18 dataset and dataloaders as well as a smoke test recipe that includes optimal parameters for training MUSDB18 models.

This dataset is a variant of the [FixedSourcesTrackFolderDataset](https://github.com/sigsep/open-unmix-pytorch/blob/master/data.py#L414) from the Open-Unmix repository.

* it was stripped down to the most basic use cases
* dataloading is limited to soundfile to not increase the number of dependencies on asteroid

This means the dataset is only compatible with MUSDB18HQ and not work with the stems version of MUSDB18. Documentation and scripts to convert MUSDB18 to wav will be added later.

### CODE REVIEW

**Comment by faroit** [comment]
@mpariente for now, please have a look at the overall structure and the API. I will iterate on that then and will polish before it can "officially" be reviewed

File: asteroid/data/musdb18_dataset.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                 source_files=['vocals', 'bass', 'drums', 'other'],
+                 suffix='.wav',
+                 seq_duration=None,
+                 samples_per_track=1,
+                 random_chunks=False,
+                 random_track_mix=False,
+                 source_augmentations=lambda audio: audio,
+                 sample_rate=44100):
+
+        """MUSDB18 torch.data.Dataset that samples from the MUSDB18 tracks
```
Comment (mpariente) [inline asteroid/data/musdb18_dataset.py:43]: We follow Google-style docstring style. 
And this should also go in the class docstring, before the `__init__` please

File: asteroid/data/musdb18_dataset.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        root : str
+            root path of MUSDB18HQ, please download from zenodo
+        suffix : str
+            specify the filename suffice, defaults to `.wav`
+        split : str
+            select dataset subfolder, defaults to ``train``.
+        subsets : list(str)
+            selects a list of track folders, defaults to `None` 
+            which loads all tracks
+        seq_duration : float
```
Comment (mpariente) [inline asteroid/data/musdb18_dataset.py:57]: We use `segment` in the other datasets? Can we change it for consistency?

File: asteroid/data/musdb18_dataset.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            # optionally select a random track for each source
+            if self.random_track_mix:
+                track_id = random.choice(range(len(self.tracks)))
+            else:
+                track_id = index // self.samples_per_track
+
+            track_path = self.tracks[track_id]['path']
+            if self.random_chunks:
+                min_duration = self.tracks[track_id]['min_duration']
+                start = random.uniform(0, min_duration - self.seq_duration)
```
Comment (mpariente) [inline asteroid/data/musdb18_dataset.py:103]: I didn't run the code but it looks like start is not defined from `random_chunks=False` (default)

File: asteroid/data/musdb18_dataset.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            specify the filename suffice, defaults to `.wav`
+        split : str
+            select dataset subfolder, defaults to ``train``.
+        subsets : list(str)
+            selects a list of track folders, defaults to `None` 
+            which loads all tracks
+        seq_duration : float
+            training is performed in chunks of ``seq_duration`` (in seconds,
+            defaults to ``None`` which loads the full audio track
+        samples_per_track : int
```
Comment (mpariente) [inline asteroid/data/musdb18_dataset.py:60]: I don't understand the goal of this argument. 
The number of epochs is here for that right? 
Or maybe I don't understand

File: asteroid/data/musdb18_dataset.py (hunk 3 above)
Comment (mpariente) [reply asteroid/data/musdb18_dataset.py:103]: Also, the start from different instruments will be different right? Is this intended?

File: egs/musdb18/OpenUnmix/test_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
+from asteroid.data import MUSDB18Dataset
+import argparse
+import torch
+from pathlib import Path
+import tqdm
+
+
+class Compose(object):
```
Comment (mpariente) [inline egs/musdb18/OpenUnmix/test_dataloader.py:8]: I've been thinking about adding this for a while.. We'll probably move it inside asteroid.

File: egs/musdb18/OpenUnmix/test_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    args = parser.parse_args()
+
+    dataset_kwargs = {
+        'root': Path(args.root),
+    }
+
+    source_augmentations = Compose(
+        [globals()['_augment_' + aug] for aug in args.source_augmentations]
+    )
```
Comment (mpariente) [inline egs/musdb18/OpenUnmix/test_dataloader.py:76]: I like that. We would probably opt for a `get` function for augmentations, as we have for norms, filterbanks etc..

File: asteroid/data/musdb18_dataset.py (hunk 3 above)
Comment (mpariente) [reply asteroid/data/musdb18_dataset.py:103]: And `min_duration` is None if `seq_duration` is None, but `random_chunks` can still be True.

**Comment by mpariente** [comment]
The overall structure is perfect! As we discussed, we can search the files only because the DB is very small, but this in fine in this case. 

We might want to integrate the transforms in the source code, but this can be done in a follow-up PR.

File: asteroid/data/musdb18_dataset.py (hunk 2 above)
Comment (faroit) [reply asteroid/data/musdb18_dataset.py:57]: sure, will adopt

File: asteroid/data/musdb18_dataset.py (hunk 1 above)
Comment (faroit) [reply asteroid/data/musdb18_dataset.py:43]: right

File: asteroid/data/musdb18_dataset.py (hunk 4 above)
Comment (faroit) [reply asteroid/data/musdb18_dataset.py:60]: as long as this is not a `IterableDataset` this param will artificially extend one epoch. Without, one epoch would be quite short (100 tracks//batch_size).

File: egs/musdb18/OpenUnmix/test_dataloader.py (hunk 5 above)
Comment (faroit) [reply egs/musdb18/OpenUnmix/test_dataloader.py:8]: yes, please

File: asteroid/data/musdb18_dataset.py (hunk 3 above)
Comment (faroit) [reply asteroid/data/musdb18_dataset.py:103]: > I didn't run the code but it looks like start is not defined from random_chunks=False (default)

good catch

Comment (faroit) [reply asteroid/data/musdb18_dataset.py:103]: > Also, the start from different instruments will be different right? Is this intended?

yes, it helps a lot. I might want to make this optional, indeed

Comment (faroit) [reply asteroid/data/musdb18_dataset.py]: | made this behavior optional and fixed the other issues

File: egs/musdb18/OpenUnmix/test_dataloader.py (hunk 6 above)
Comment (faroit) [reply egs/musdb18/OpenUnmix/test_dataloader.py:76]: I would change this later when you've added the Compose class to asteroid core

File: asteroid/data/musdb18_dataset.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        # load sources
+        for source in self.sources:
+            # optionally select a random track for each source
+            if self.random_track_mix:
+                # load a different track
+                track_id = random.choice(range(len(self.tracks)))
+                if self.random_segments:
+                    start = random.uniform(
+                        0, self.tracks[track_id]['min_duration'] - self.segment
+                    )
```
Comment (mpariente) [inline asteroid/data/musdb18_dataset.py:135]: This makes sense to me, at least, when we don't mix track together, we keep things aligned

File: asteroid/data/musdb18_dataset.py (hunk 4 above)
Comment (mpariente) [reply asteroid/data/musdb18_dataset.py]: Ok. I don't see the use case but we can keep it for now.

**Comment by mpariente** [comment]
Looks great, thank you very much ! 
Cannot wait for the Open-Unmix recipe  :smiley:

**Comment by faroit** [comment]
>  Cannot wait for the Open-Unmix recipe 😃

actually what about we start with pulling the model using `torch.hub.load('sigsep/open-unmix-pytorch', ...)` and train that in the recipe until there is a need to diverge from our sigsep repo? That would reduce confusion about the asteroid/openunmix vs sigsep/openunmix. And it would still allow us to use asteroid loss functions and asteroid optimizers/tricks...

**Comment by mpariente** [comment]
Oh yes, I didn't think about that actually, it's much better !

**Comment by faroit** [comment]
> Oh yes, I didn't think about that actually, it's much better !

Great, I can do that pretty quickly. I was about to update the training code to lighting anyway

**Comment by faroit** [comment]
@mpariente we forgot to invite @adefossez to review the MUSDB dataset code. @adefossez please review this with respect to mean/std properties you need for demucs and lets add another PR if necessary

**Comment by mpariente** [comment]
I assumed you converged beforehand. Yes, @adefossez feel free to review.

Also : addresses part of #47 .

**Comment by adefossez** [comment]
Hey guys, sorry for replying quite late on this. Indeed I was doing mean/std normalization for training Demucs, at the track level, not segment level, but to be honest, I don't think this is really worth adding here, because it either means loading the entire dataset to compute it every training, or storing some state somewhere which I don't like (what if one of the file changes, etc.)
I think I would rather change slightly the normalization so that it can be performed on the fly, using a moving average of a given window size, with the window being equal or smaller to the training segment duration.
